servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublishAndSubscribeOnCompletables.java

```diff
@@ -110,8 +107,11 @@ void handleSubscribe(final Subscriber subscriber,
                 Subscriber wrapped = contextProvider.wrapCancellable(subscriber, contextMap);
 
                 // offload the remainder of subscribe()
-                LOGGER.trace("Offloading Completable subscribe() on {}", executor());
-                executor().execute(() -> super.handleSubscribe(wrapped, contextMap, contextProvider));
+                if (shouldOffload.getAsBoolean()) {
```

Comment:
`public final Publisher<T> subscribeOn(Executor executor, BooleanSupplier shouldOffload)` -> `public final Publisher<T> subscribeOn(Executor executor, Supplier<? extends BooleanSupplier> shouldOffload)`

inlined with our discussion yesterday I think we can change the operator signature to a `Supplier` and just override `handleSubscribe` such that you don't have to call `apply`:

```java
private static final class SubscribeOn<T> extends AbstractNoHandleSubscribePublisher<T> {
  private final Supplier<? extends BooleanSupplier> shouldOffloadSupplier; // set in constructor

  @Override
  final void handleSubscribe(Subscriber<? super R> subscriber,
                               AsyncContextMap contextMap, AsyncContextProvider contextProvider) {
     BooleanSupplier shouldOffload = shouldOffloadSupplier.get();

     // Note that the Executor is wrapped by default to preserve async context, so if the executor is
     // wrapped we don't have to re-wrap the Subscriber. Also we only "need" to preserve async context
     // when offloading happens, as otherwise this operator is synchronous. add tests for this :)
     Subscriber<? super T> wrapped = new OffloadedSubscriptionSubscriber<>(subscriber, shouldOffload, executor());
     if (shouldOffload.getAsBoolean()) {
       executor().execute(() -> super.handleSubscribe(wrapped, contextMap, contextProvider));
     } else {
       super.handleSubscribe(wrapped, contextMap, contextProvider);
     }
  }
}
```